## 1. Importing libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

## 2. Importing Data

In [2]:
# Turn project folder path into a string
path = r'C:\Users\Admin\Desktop\Data Analysis\Python\04-2024 Instacart Basket Analysis'

In [3]:
path

'C:\\Users\\Admin\\Desktop\\Data Analysis\\Python\\04-2024 Instacart Basket Analysis'

In [4]:
# Import orders_products_merge.pkl
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merge.pkl'))

In [5]:
# Checking the shape of ords_prods_merge
ords_prods_merge.shape

(32404289, 14)

In [6]:
# checking the output of ords_prods_merge
ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge  
0         Soda        77              7     9.0   both  
1         Soda        77              7     9.0   both  
2         Soda        77              7     9.0   both  
3         Soda        77              7     9.0   both  
4         Soda        77              7     9.0   both

In [7]:
# Droping the _merge columns
ords_prods_merge = ords_prods_merge.drop(columns = ['_merge'])

In [8]:
# Checking the output of df_ords_prods
ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices  
0         Soda        77              7     9.0  
1         Soda        77              7     9.0  
2         Soda        77              7     9.0  
3         Soda        77              7     9.0  
4         Soda        77              7     9.0

In [9]:
# Checking the shape of df_ords_prods after drops
ords_prods_merge.shape

(32404289, 13)

In [10]:
# Create a subset with only first 1 Million lines for ords_prods_merge
df = ords_prods_merge[:1000000]

In [11]:
# Checking the shape of df
df.shape

(1000000, 13)

In [12]:
# Checking the output of df
df.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices  
0         Soda        77              7     9.0  
1         Soda        77              7     9.0  
2         Soda        77              7     9.0  
3         Soda        77              7     9.0  
4         Soda        77              7     9.0

## 3. Price Label Function (With IF-ELSE)

In [13]:
# Function to designating row as a low-, mid-, or high-range product based on its price
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [14]:
# Apply the price_label function
df['price_range'] = df.apply(price_label, axis = 1)

C:\Users\Admin\AppData\Local\Temp\ipykernel_9416\3441332910.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis = 1)


In [15]:
# Check the values of the new column 'price_range'
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    756450
Low-range product    243550
Name: count, dtype: int64

In [16]:
# Checking for max price in the df dataset
df['prices'].max()

14.8

## 4. Labeling Price with .loc

In [26]:
# Assignging 'High-range product' when price > 15 in 'price_range_loc' column
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [18]:
# Assigning 'Mid-range product' when price is between 5 and 15 in 'price_range_loc' column
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [19]:
# Assigning 'Low-range product' when price is less than 5 in 'price_range_loc' column
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [20]:
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product    756450
Low-range product    243550
Name: count, dtype: int64

## 5. For-Loops

In [21]:
# Checking the frequency of orders in a week
ords_prods_merge['order_day_of_week'].value_counts(dropna = False)

order_day_of_week
0    6204040
1    5660135
6    4496403
2    4213760
5    4205721
3    3840476
4    3783754
Name: count, dtype: int64

In [22]:
# For loop to assign 0 as the busiest day, 4 as the least busy day, everything else as regularly busy
result = []

for value in ords_prods_merge["order_day_of_week"]:
    if value == 0:
        result.append("Busiest day")
    elif value == 4:
        result.append("Least busy")
    else:
        result.append("Regularly busy")

In [23]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [24]:
# Assigning result list to a new column in ords_prods_merge
ords_prods_merge['busiest_day'] = result

In [25]:
ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22416495
Busiest day        6204040
Least busy         3783754
Name: count, dtype: int64

In [28]:
df.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices        price_range  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

     price_range_loc  
0  Mid-range product  
1  Mid-range product  
2  Mid-range product  
3  Mid-range product  
4  Mid-range product

## 6.Task

### 6.2 Suppose your clients have changed their minds about the labels you created in your “busiest_day” column. Now, they want “Busiest day” to become “Busiest days” (plural). This label should correspond with the two busiest days of the week as opposed to the single busiest day. At the same time, they’d also like to know the two slowest days. Create a new column for this using a suitable method.

In [29]:
# Checking the frequency of orders in a week
ords_prods_merge['order_day_of_week'].value_counts(dropna = False)

order_day_of_week
0    6204040
1    5660135
6    4496403
2    4213760
5    4205721
3    3840476
4    3783754
Name: count, dtype: int64

In [32]:
# For loop to assign 0 & 1 as the busiest days, 3 & 4 as the slowest days, everything else as regularly busy
result_new = []

for value in ords_prods_merge["order_day_of_week"]:
    if value == 0 or value == 1:
        result_new.append("Busiest days")
    elif value == 3 or value == 4:
        result_new.append("Slowest days")
    else:
        result_new.append("Regularly busy")

In [33]:
result_new

['Regularly busy',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Slowest days',
 'Regularly busy',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Slowest days',
 'Slowest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest 

In [34]:
# Assigning result list to a new column in ords_prods_merge
ords_prods_merge['busiest_day_new'] = result_new

### 6.3 Check the values of this new column for accuracy. Note any observations in markdown format.

In [36]:
# Checking the results
ords_prods_merge['busiest_day_new'].value_counts(dropna = False)

busiest_day_new
Regularly busy    12915884
Busiest days      11864175
Slowest days       7624230
Name: count, dtype: int64

#### Row enteries tally up to 32,404,289 for 'busiest_day' and 'busiest_day_new', so all rows are accounted for.

### 6.4 When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”

In [37]:
# Checking the frequency of orders in hours of day
ords_prods_merge['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    2761710
11    2736075
14    2689086
15    2662094
13    2660900
12    2618481
16    2535154
9     2454165
17    2087609
8     1718100
18    1636469
19    1258290
20     976145
7      891040
21     795628
22     634216
23     402315
6      290492
0      218766
1      115699
5       87959
2       69374
4       53241
3       51281
Name: count, dtype: int64

In [38]:
# Section the hours into 3 sections, as follows:
# Most orders: 9, 10, 11, 12, 13, 14, 15, 16
# Fewest orders: 23, 0, 1, 2, 3, 4, 5, 6
# Average orders: every other hour not in most order and fewest order.
# For loop to assign the above
result_hour = []

for value in ords_prods_merge["order_hour_of_day"]:
    if value == 9 or value == 10 or value == 11 or value == 12 or value == 13 or value == 14 or value == 15 or value == 16:
        result_hour.append("Most orders")
    elif value == 23 or value == 0 or value == 1 or value == 2 or value == 3 or value == 4 or value == 5 or value == 6:
        result_hour.append("Fewest orders")
    else:
        result_hour.append("Average orders")

In [40]:
result_hour

['Average orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most ord

In [41]:
# Assigning result list to a new column in ords_prods_merge
ords_prods_merge['busiest_period_of_day'] = result_hour

### 6.5 Print the frequency for this new column.

In [42]:
# Checking the results
ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders       21117665
Average orders     9997497
Fewest orders      1289127
Name: count, dtype: int64

### 6.6 Export dataframe

In [44]:
# Export ords_prods_merge to pkl
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_updated.pkl'))